### PR TITLE
C++: instruction -> operand field flow

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR changes `readStep` so that it performs `instruction -> operand` flow, and `storeStep` so that it performs `operand -> instruction` flow. This matches better with the way we do simple flow steps.

CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1527/

### CODE REVIEW

**Comment by MathiasVP** [comment]
The failing tests all look like changes in path explanation in the internal repo. Once the changes here have been approved I'll make an internal PR.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:231]: Won't this start picking up inexact defs? And if so, is that good or bad?

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:255]: Same questions about partial defs.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:401]:
```suggestion
    address = skipCopyValueInstructions(operand.getAddressOperand()) and
```
There are actually three different instructions that have `LoadOperand`s: `Load`, `ReturnValue`, and `ThrowValue`. It's probably good that this predicate now handles all three, but instead of relying on the use to be a `LoadInstruction`, you should use the simpler and more general `MemoryOperand::getAddressOperand()`.

**Review by dbartol — CHANGES_REQUESTED** [review]
A few questions and a few comments.

Would you mind waiting to merge this until the temporary object PR is merged? I don't think we have any merge conflicts quite yet, but we'll probably conflict on those path explanation tests in the private repo, and your PR may provoke real analysis change in mine.

**Comment by MathiasVP** [comment]
> Would you mind waiting to merge this until the temporary object PR is merged? I don't think we have any merge conflicts quite yet, but we'll probably conflict on those path explanation tests in the private repo, and your PR may provoke real analysis change in mine.

That's a good point. I'll let this PR wait until your temporary object PR is merged.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:231]: Good point. I didn't actually think of that. For read steps we allow inexact defs (as otherwise we wouldn't pick up a load of a field address). What source code would generate a store instruction with a non-exact source value operand anyway?

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:255]: You're right. It shouldn't actually be a `StoreValueOperand`, but rather a `ChiPartialOperand`. From `simpleOperandLocalFlowStep` we get flow from the result of the store instruction to any total exact overlapping use. So if we add a `storeStep` from the partial chi operand to its chi instruction we get `store -> chi partial operand -> chi instruction` flow.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:401]: Cool. I wasn't aware of that predicate. That's really useful.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:273]: And pretty much the same answer :) It should be a `ChiPartialOperand`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Fixed in 3c778868f8a5142a6f1862167226b0609f7c1658. I had to modify the definition of `skipCopyValueInstructions` to take an `Operand` as an argument as there was no `getAddressOperandDef()`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Fixed in 3c778868f8a5142a6f1862167226b0609f7c1658.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:256]: Fixed in 3c778868f8a5142a6f1862167226b0609f7c1658. Surprisingly, it had zero effect on the path explanations in the library-test files (but a few in the query-test files).

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:231]: It's a register operand, so I don't think it can be inexact.

**Comment by MathiasVP** [comment]
> CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1527/

Performance looks alright. The only worry is a 4% analysis-time increase on `boost`, but it's not something I can reproduce locally, and there's nothing suspicious in the evaluation log.

**Comment by MathiasVP** [comment]
Now that https://github.com/github/codeql/pull/4432 has been merged I think this PR is good to go. I've created an internal PR for the test changes that are making Actions to complain.

**Review by dbartol — APPROVED** [review]
Thanks for the fixes. LGTM.

**Pull request merged**
